### PR TITLE
Multithreaded indexing through Elasticsearch (with ES 7 & Lucene 8)

### PR DESCRIPTION
**Pull request by charW**
This is part of the effort of integrating Elasticsearch into Anserini. 

**Update:** we have recently rebased the branch `add-elasticsearch-lucene7` so it shares the same Lucene 8 dependency as the master branch (this makes the branch name a misnomer), and we have also updated this branch to use Elasticsearch 7.0.0.

### CODE REVIEW

**Comment by r-clancy** [comment]
When documents are indexed, they are placed in an index based on the name of the file they came from. Can we add a `-es.index` option (just like `-solr.index`) to specify the name of the index?
image

**Comment by r-clancy** [comment]
Currently, both the `raw` and `contents` fields are indexed into ES regardless of the values of `-storeRawDocs` (which could be raw HTML) and `-storeTransforedDocs` (which could be parsed text from HTML), respectively. This isn't an issue in the Solr version (which the ES one follows) as it's done via configuration and not code in Solr.

When the `-storeRawDocs` flag is passed, `raw` should be indexed and when the `-storeTransformedDocs` flag is passed, `contents` should be stored.

File: src/main/java/io/anserini/index/IndexCollection.java [hunk 1]
```diff
@@ -400,6 +449,121 @@ private void flush() {
...
+          builder.startObject();
+          for (IndexableField field : document.getFields()) {
+            if (field.stringValue() != null) {
+              builder.field(field.name(), field.stringValue());
+            } else if (field.numericValue() != null) {
+              builder.field(field.name(), field.numericValue());
+            }
+          }
+          builder.endObject();
+          bulkRequest.add(new IndexRequest(input.getFileName().toString(), "doc", sourceDocument.id()).source(builder));
```
Comment (r-clancy) [inline src/main/java/io/anserini/index/IndexCollection.java]: When documents are indexed, they are placed in an index based on the name of the file they came from. Can we add a -es.index option (just like -solr.index) to specify the name of the index?

image

File: src/main/java/io/anserini/index/IndexCollection.java [hunk 2]
```diff
@@ -400,6 +449,121 @@ private void flush() {
...
+            continue;
+          }
+          if (whitelistDocids != null && !whitelistDocids.contains(sourceDocument.id())) {
+            counters.skipped.incrementAndGet();
+            continue;
+          }
+
+          XContentBuilder builder = XContentFactory.jsonBuilder();
+          builder.startObject();
+          for (IndexableField field : document.getFields()) {
```
Comment (r-clancy) [inline src/main/java/io/anserini/index/IndexCollection.java]: Currently, both the raw and contents fields are indexed into ES regardless of the values of -storeRawDocs (which could be raw HTML) and -storeTransforedDocs (which could be parsed text from HTML), respectively. This isn't an issue in the Solr version (which the ES one follows) as it's done via configuration and not code in Solr.

When the -storeRawDocs flag is passed, raw should be indexed and when the -storeTransformedDocs flag is passed, contents should be stored.

**Comment by r-clancy** [comment]
The `-storeRawDocs` and `-storeTransformedDocs` changes look good!

A couple things I noticed...

The performance of ES is quite slow in comparison to Lucene (expected), but also Solr - take a look into this? From a quick search, there appears to be a `index.refresh_interval` option, which controls how quickly documents are available for searching, that defaults to 1 second. In the Solr implementation, we control similar behavior with the `solr.commitWithin` option. Let's have some more fine grained control over when documents are committed within ES? I'm not sure what the `RequestOptions` mechanism exposes, if not that maybe update the refresh interval for the index to `60s` by default like Solr?

Once indexing is complete, the program doesn't appear to exit and I need to Ctrl-C to get out of it. I'll try on another PC just to make sure, but maybe some resource isn't being closed properly? I don't see the ES client pool being closed...

**Comment by r-clancy** [comment]
@charW Can we also rebase this branch to the current `master` which is now Lucene 8.0.0?

**Comment by r-clancy** [comment]
@emmileaf @lintool Can one of you follow the `elastirini.md` and sanity check this PR before we merge?

File: docs/elastirini.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+`curl --user elastic:changeme -XPUT "localhost:9200/yourindexname"`
+
+Here, the username and password are those defaulted by `docker-elk`. You can change these if you like.
+
+You can further specify the settings associated with this index. For example, if you would like to change `index.refresh_interval` from the default 1 second to 60 seconds:
+
+`curl --user elastic:changeme -XPUT -H 'Content-Type: application/json' 'localhost:9200/yourindexname/_settings' -d '{ "index": { "refresh_interval": "60s"}}'`
+
+Now, we can start indexing through Elastirini. Here, instead of passing in `-index` (to index with Lucene directly) or `-solr` (to index with Solr), we pass in `-es`. For example, to index [robust04](https://github.com/castorini/Anserini/blob/master/docs/experiments-robust04.md), we could run:
```
Comment (r-clancy) [inline docs/elastirini.md]: URL has changed to https://github.com/castorini/anserini/blob/master/docs/regressions-robust04.md

**Comment by emmileaf** [comment]
> @emmileaf @lintool Can one of you follow the `elastirini.md` and sanity check this PR before we merge?

Was able to setup and index `robust04` according to doc. LGTM!

**Pull request merged**